libraries/AP_Motors/AP_MotorsHeli_Dual.cpp

```diff
@@ -203,35 +203,25 @@ void AP_MotorsHeli_Dual::output_test_seq(uint8_t motor_seq, int16_t pwm)
 // set_desired_rotor_speed
 void AP_MotorsHeli_Dual::set_desired_rotor_speed(float desired_speed)
 {
-    _rotor.set_desired_speed(desired_speed);
+    _main_rotor.set_desired_speed(desired_speed);
 }
 
 // set_rotor_rpm - used for governor with speed sensor
 void AP_MotorsHeli_Dual::set_rpm(float rotor_rpm)
 {
-    _rotor.set_rotor_rpm(rotor_rpm);
+    _main_rotor.set_rotor_rpm(rotor_rpm);
 }
 
 // calculate_armed_scalars
 void AP_MotorsHeli_Dual::calculate_armed_scalars()
 {
-    // Set common RSC variables
-    _rotor.set_ramp_time(_rsc_ramp_time);
-    _rotor.set_runup_time(_rsc_runup_time);
-    _rotor.set_critical_speed(_rsc_critical*0.001f);
-    _rotor.set_idle_output(_rsc_idle_output*0.001f);
-    _rotor.set_slewrate(_rsc_slewrate);
-
     // Set rsc mode specific parameters
-    if (_rsc_mode == ROTOR_CONTROL_MODE_OPEN_LOOP_POWER_OUTPUT) {
-        _rotor.set_throttle_curve(_rsc_thrcrv.get_thrcrv());
-    } else if (_rsc_mode == ROTOR_CONTROL_MODE_CLOSED_LOOP_POWER_OUTPUT) {
-        _rotor.set_throttle_curve(_rsc_thrcrv.get_thrcrv());
-        _rotor.set_governor_disengage(_rsc_gov.get_disengage()*0.01f);
-        _rotor.set_governor_droop_response(_rsc_gov.get_droop_response()*0.01f);
-        _rotor.set_governor_reference(_rsc_gov.get_reference());
-        _rotor.set_governor_range(_rsc_gov.get_range());
-        _rotor.set_governor_thrcurve(_rsc_gov.get_thrcurve()*0.01f);
+    if (_main_rotor._rsc_mode.get() == ROTOR_CONTROL_MODE_OPEN_LOOP_POWER_OUTPUT || _main_rotor._rsc_mode.get() == ROTOR_CONTROL_MODE_CLOSED_LOOP_POWER_OUTPUT) {
+        _main_rotor.set_throttle_curve();
+    }
+    if (_main_rotor._rsc_mode.get() != _main_rotor.get_control_mode()) {
```

Comment:
@bnsgeyer, can I ask what this check is for?  I don't really understand the relationship between _control_mode and _rsc_mode but I worry about two things happening when the vehicle is armed (I assume calculate_armed_scalars is running when the vehicle is armed):

- we could end up calling .reset_rsc_mode_param() which does a save to eeprom which (I think) can take a significant amount of time and stall the main loop
- we could swamp the telemetry links with the "RSC control mode change failed" message.

I wonder if this check could be moved to when the vehicle is disarmed or make i a pre-arm check or something.  Also I wonder a bit why it's here instead of where the _rsc_mode or _control_mode are set.  I.e. maybe it would be better to check right where one of them is set to the other rather than in calculate-armed-scalars.

Anyway, I don't understand fully but it seems a bit odd so I wanted to check with you before we merge.

Reply:
@rmackay9 I put this here to keep users from changing the RSC control mode while flying.  RSC mode and control mode are the same thing.  RSC mode is the parameter and control mode is the actual RSC mode that is currently being used in the RSC.  Previously the RSC mode parameter would be changed but the control mode would not which is the desired response but the Setpoint would change based on the RSC mode parameter.  So a lot of undesirable stuff going on.  this is designed to refuse the mode change while armed and keep the setpoint from changing based on the RSC mode parameter. Well I looked back and found one other change that would guarantee that the setpoint would not change with the RSC mode parameter.  In Heli.cpp, I need to have it get the current control mode and not the rsc mode parameter.  that would fix the issue I was seeing and I could get away with not doing the param reset.  However it would change the rsc mode once the aircraft was disarmed.  Not sure if that is good or bad.  If the rsc mode parameter is set while armed, wouldn't this statement only be called once? Or are you mainly concerned with the save portion of the reset.  Not sure how I could track that and just set and save after disarm.  I'm open to suggestions.